### PR TITLE
Changing &var[0] to var.data()

### PR DESCRIPTION
**Pull request by meshcollider**
This just continues the work of https://github.com/bitcoin/bitcoin/pull/9804

Modifies a lot of `&vector[]`'s to `vector.data()`'s across all the files including tests, just the stuff that 9804 missed

### CODE REVIEW

**Comment by promag** [comment]
#9804 should cherry pick this one right?

**Comment by MeshCollider** [comment]
#9804 has just been merged

File: src/rpc/server.cpp [hunk 1]
```diff
@@ -444,7 +444,7 @@ static inline JSONRPCRequest transformNamedArguments(const JSONRPCRequest& in, c
     const std::vector<UniValue>& values = in.params.getValues();
     std::unordered_map<std::string, const UniValue*> argsIn;
     for (size_t i=0; i<keys.size(); ++i) {
-        argsIn[keys[i]] = &values[i];
+        argsIn[keys[i]] = values.data() + i;
```
Comment (jtimon) [inline src/rpc/server.cpp]: Is this really clearer?

File: src/test/bloom_tests.cpp [hunk 2]
```diff
@@ -154,8 +154,8 @@ BOOST_AUTO_TEST_CASE(bloom_match)
     COutPoint prevOutPoint(uint256S("0x90c122d70786e899529d71dbeba91ba216982fb6ba58f3bdaab65e73b7e9260b"), 0);
     {
         std::vector<unsigned char> data(32 + sizeof(unsigned int));
-        memcpy(&data[0], prevOutPoint.hash.begin(), 32);
-        memcpy(&data[32], &prevOutPoint.n, sizeof(unsigned int));
+        memcpy(data.data(), prevOutPoint.hash.begin(), 32);
+        memcpy(data.data()+32, &prevOutPoint.n, sizeof(unsigned int));
```
Comment (jtimon) [inline src/test/bloom_tests.cpp]: same here

File: src/test/skiplist_tests.cpp [hunk 3]
```diff
@@ -20,13 +20,13 @@ BOOST_AUTO_TEST_CASE(skiplist_test)
 
     for (int i=0; i<SKIPLIST_LENGTH; i++) {
         vIndex[i].nHeight = i;
-        vIndex[i].pprev = (i == 0) ? NULL : &vIndex[i - 1];
+        vIndex[i].pprev = (i == 0) ? NULL : (vIndex.data() + i - 1);
```
Comment (jtimon) [inline src/test/skiplist_tests.cpp]: same here

File: src/test/skiplist_tests.cpp [hunk 4]
```diff
@@ -20,13 +20,13 @@ BOOST_AUTO_TEST_CASE(skiplist_test)
...
         vIndex[i].BuildSkip();
     }
 
     for (int i=0; i<SKIPLIST_LENGTH; i++) {
         if (i > 0) {
-            BOOST_CHECK(vIndex[i].pskip == &vIndex[vIndex[i].pskip->nHeight]);
+            BOOST_CHECK(vIndex[i].pskip == vIndex.data() + vIndex[i].pskip->nHeight);
```
Comment (jtimon) [inline src/test/skiplist_tests.cpp]: here you aren't even consistent

File: src/test/skiplist_tests.cpp [hunk 5]
```diff
@@ -37,9 +37,9 @@ BOOST_AUTO_TEST_CASE(skiplist_test)
         int from = InsecureRandRange(SKIPLIST_LENGTH - 1);
         int to = InsecureRandRange(from + 1);
 
-        BOOST_CHECK(vIndex[SKIPLIST_LENGTH - 1].GetAncestor(from) == &vIndex[from]);
-        BOOST_CHECK(vIndex[from].GetAncestor(to) == &vIndex[to]);
-        BOOST_CHECK(vIndex[from].GetAncestor(0) == &vIndex[0]);
+        BOOST_CHECK(vIndex[SKIPLIST_LENGTH - 1].GetAncestor(from) == vIndex.data() + from);
```
Comment (jtimon) [inline src/test/skiplist_tests.cpp]: well, you get the idea, if the index was other than 0, it seems to me that the change makes things worse.

**Comment by jtimon** [comment]
Fast review ACK modulo nits.

File: src/test/skiplist_tests.cpp (hunk 4 above)
Comment (dcousens) [inline src/test/skiplist_tests.cpp]: what lacks consistency?

Comment (jtimon) [inline src/test/skiplist_tests.cpp]: One time vIndex is accessed by ```vIndex.data() + n``` and another with ```vIndex[m]```

Comment (MeshCollider) [inline src/test/skiplist_tests.cpp]: That's because one is accessing the element not a pointer...

**Comment by MeshCollider** [comment]
@jtimon this is consistent with how #9804 did it

File: src/test/skiplist_tests.cpp (hunk 4 above)
Comment (dcousens) [inline src/test/skiplist_tests.cpp]: The `.data()` usage here however,  is not really inline with #9804.
#9804 replaced usage of `&v[0]` and `&v[N]` for specifying a slice/range.
In this change above,  you are specifically referring to the memory location of 1 element,  which isn't an rvalue... probably unnecessary to use `.data()`.

File: src/test/bloom_tests.cpp (hunk 2 above)
Comment (dcousens) [inline src/test/bloom_tests.cpp]: IMHO this is clearer

File: src/test/skiplist_tests.cpp (hunk 4 above)
Comment (MeshCollider) [inline src/test/skiplist_tests.cpp]: Isn't it still preferable to access the memory location of an element this way?

Comment (dcousens) [inline src/test/skiplist_tests.cpp]: @MeshCollider the advantage of `.data()` is that it is defined to return `nullptr` if `.empty()`,  compared to `&v[0]` potentially de-referencing unitialized memory.
In this event,  we assume `vIndex` is not empty, as you are de-referencing `vIndex[i]`,  therefore, it is probably not as necessary.

**Comment by MeshCollider** [comment]
Reverted the changes where specific elements were accessed, thanks @jtimon and @dcousens

**Comment by MeshCollider** [comment]
Rebased

**Comment by MeshCollider** [comment]
Rebased

File: src/qt/test/paymentservertests.cpp [hunk 6]
```diff
@@ -43,7 +43,7 @@ static SendCoinsRecipient handleRequest(PaymentServer* server, std::vector<unsig
     // Write data to a temp file:
     QTemporaryFile f;
     f.open();
-    f.write((const char*)&data[0], data.size());
+    f.write((const char*) data.data(), data.size());
```
Comment (laanwj) [inline src/qt/test/paymentservertests.cpp]: nit: Please remove space after the case here

**Comment by laanwj** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/10793/commits/db0e991fb7975c071818b058376644be08ced343

File: src/qt/test/paymentservertests.cpp (hunk 6 above)
Comment (MarcoFalke) [reply src/qt/test/paymentservertests.cpp]: @MeshCollider You can pipe the diff through clang-format to prevent such nits in the future. See dev notes.

Comment (MeshCollider) [reply src/qt/test/paymentservertests.cpp]: Fixed, thanks :)

File: src/test/bloom_tests.cpp (hunk 2 above)
Comment (laanwj) [reply src/test/bloom_tests.cpp]: If data is empty, I'd say copying to data()+32 is U/B no matter what.

File: src/rpc/server.cpp (hunk 1 above)
Comment (laanwj) [reply src/rpc/server.cpp]: Agree with @jtimon here. This PR is titled `Changing &var[0] to var.data()`, this seems to overshoot that scope. Addressing out of an vector is UB no matter what, so this doesn't fix any potential issues, and isn't more readable IMO either.

**Comment by laanwj** [comment]
The commit message is weird now :)

**Comment by MeshCollider** [comment]
Argh fixed, thanks :)

**Pull request merged**